sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala

```diff
@@ -122,24 +119,22 @@ case class MemoryStream[A : Encoder](id: Int, sqlContext: SQLContext)
 
   override def toString: String = s"MemoryStream[${Utils.truncatedString(output, ",")}]"
 
-  override def setOffsetRange(start: Optional[OffsetV2], end: Optional[OffsetV2]): Unit = {
-    synchronized {
-      startOffset = start.orElse(LongOffset(-1)).asInstanceOf[LongOffset]
-      endOffset = end.orElse(currentOffset).asInstanceOf[LongOffset]
-    }
-  }
-
   override def deserializeOffset(json: String): OffsetV2 = LongOffset(json.toLong)
 
-  override def getStartOffset: OffsetV2 = synchronized {
-    if (startOffset.offset == -1) null else startOffset
+  override def initialOffset: OffsetV2 = LongOffset(-1)
+
+  override def latestOffset(start: OffsetV2): OffsetV2 = {
+    if (currentOffset.offset == -1) null else currentOffset
```

Comment:
I feel it's more reasonable to forbid data source to return null offsets, and use `latestOffset != startOffset` at the streaming execution side to indicate if there is a new batch available. But I'll leave it to followup PR.

cc @jose-torres

Reply:
Yes, I agree. The V1 API didn't require sources to implement a "this is the beginning of the stream, read everything" offset, but that was a mistake we should make sure to remedy here.

A followup PR makes sense, because there's some stream execution logic that can be greatly simplified when all sources have a real initial offset.